### PR TITLE
Multiple ways of resolving the object's storage path

### PR DESCRIPTION
**Pull request by gnutix**
Fixes https://github.com/dustin10/VichUploaderBundle/issues/858#issuecomment-511848521

### CODE REVIEW

**Comment by gnutix** [comment]
I can't get the same error in local using `vendor/bin/phpunit`. Any idea how I can reproduce it ?

File: Form/Type/VichImageType.php [hunk 1]
```diff
@@ -19,6 +19,12 @@
  */
 class VichImageType extends VichFileType
```
Comment (garak) [inline Form/Type/VichImageType.php]: Why on VichImageType? Don't you think that such change would fit on VichFileType too?

Comment (gnutix) [reply Form/Type/VichImageType.php]: This code is meant only to generate the value of `image_uri` (which is specific to `VichImageType`, not `VichFileType`). So it feels in the right place to me.

Comment (garak) [reply Form/Type/VichImageType.php]: I guess that the problem described in related issue can apply also to a generic file: the user want to display a link to file, and link href attribute is technically the same as image src attribute (a path, in the end)

Comment (garak) [reply Form/Type/VichImageType.php]: You're right, the option applies only to VichImageType

File: Form/Type/VichImageType.php [hunk 2]
```diff
@@ -19,6 +19,12 @@
...
 {
+    const STORAGE_RESOLVE_URI = 0;
```
Comment (garak) [inline Form/Type/VichImageType.php]: can we make constants private?

File: Form/Type/VichImageType.php [hunk 3]
```diff
@@ -90,4 +106,21 @@ public function getBlockPrefix(): string
     {
         return 'vich_image';
     }
+
+    protected function resolvePath($storageResolveMethod, $object, FormInterface $form)
```
Comment (garak) [inline Form/Type/VichImageType.php]: Can we make this method private?
Can we add a return type?

File: Tests/Form/Type/VichImageTypeTest.php [hunk 4]
```diff
@@ -204,6 +213,45 @@ public function testLiipImagineBundleIntegration(): void
         $this->assertEquals($vars, $view->vars);
     }
 
+    public function getLiipImagineBundleIntegrationData(): array
+    {
+        $field = 'image';
+        $object = new Product();
+
+        return [
+            [
```
Comment (garak) [inline Tests/Form/Type/VichImageTypeTest.php]: can we use keys to describe provided cases?

File: Form/Type/VichImageType.php (hunk 2 above)
Comment (gnutix) [reply Form/Type/VichImageType.php]: Sure! I was not sure about the PHP supported version, but as it's 7.2 it's all good.

Comment (gnutix) [reply Form/Type/VichImageType.php]: Having them public allows to make the configuration quite readable :

```php
$builder->add('file', VichImageType::class, [
    'storage_resolve_method' => VichImageType::STORAGE_RESOLVE_PATH_ABSOLUTE,
]);
```

Where having them private would force to use integers or pre-defined strings for configuration. Not as nice IMO.

File: Tests/Form/Type/VichImageTypeTest.php (hunk 4 above)
Comment (gnutix) [reply Tests/Form/Type/VichImageTypeTest.php]: Done

File: Form/Type/VichImageType.php (hunk 3 above)
Comment (gnutix) [reply Form/Type/VichImageType.php]: Done

**Comment by gnutix** [comment]
Hmm... I just realized there's another way to accomplish what I needed : 

```php
<?php
use Liip\ImagineBundle\Imagine\Cache\CacheManager;

final class MyFormType extends AbstractType
{
    private $cacheManager;

    public function __construct(CacheManager $cacheManager)
    {
        $this->cacheManager = $cacheManager;
    }

    public function buildForm(FormBuilderInterface $builder, array $options): void
    {
        $builder->add('file', VichImageType::class, [
            'image_uri' => function ($object, $resolvedUri) {
                return $this->cacheManager->getBrowserPath($object->getFileName(), 'my_imagine_filter');
            },
        ]);
    }
}
```

That way, I bypass the "storage->resolveUri" call and use my "relative filename" to generate the path using LiipImagine's CacheManager directly.

I'm wondering if that couldn't be the simplest answer to this issue, instead of adding all this new code. WDYT @BlackWiCKED ? Could you please try it out and let us know if that works for your case ?

**Comment by BlackWiCKED** [comment]
> I'm wondering if that couldn't be the simplest answer to this issue, instead of adding all this new code. WDYT @BlackWiCKED ? Could you please try it out and let us know if that works for your case ?

Thank you for your efforts. It works partially @gnutix, it's a good thing that I can access the cached resources this way, but it also generates a lot of extra (and not necessarily desirable) work on the user side. I have more entities, and inserting the same closure into each form type is not what I would prefer - plus I need to set the constructor function and use the CacheManager class in each of my form type classes.

I said it works partially, because I noticed some strange behavior too. It might be specific to my case only though - but it did not happen with version 1.9.2 so I think it's worth mentioning. When there is no image uploaded to my entity, I use a placeholder image - which is an asset image file in my source code. That file is not stored in the private S3 bucket, but on my own server. It still gets cached though, and that cached image from the public S3 bucket is displayed on the site the same way as regular uploaded images. Basically I have two loaders (filesystem_loader and flysystem_loader) merged together with a chain loader in liip imagine config. 

`{{ (file.image|default(asset('build/images/noimage.png')))|imagine_filter('thumbnail') }}`

Looks like the problem is that when **_file.image_** is not set, the closure is still called via "image_uri" property, and since **_$object->getImage()_** is empty, **_$this->cacheManager->getBrowserPath_** returns a non existing path - it's just the cache folder - without the filename. The missing file appears in the browser's network tab. It can be fixed with something like this:

```php
$builder->add('file', VichImageType::class, [
    'image_uri' => function ($object/*, $resolvedUri*/) {
        if (!empty($object->getImage())) {
            return $this->cacheManager->getBrowserPath($object->getImage(), 'thumbnail');
        }
        return false;
    }
]);
```

Anyways, my opinion is that it just adds more complexity to something that should not be there at all. Setting properties like "imagine_pattern" or "storage_resolve_method" is fine, but writing and duplicating actual code - just to make a 3rd party library work - is not what I would prefer.

File: Form/Type/VichImageType.php [hunk 5]
```diff
@@ -90,4 +106,21 @@ public function getBlockPrefix(): string
     {
         return 'vich_image';
     }
+
+    private function resolvePath($storageResolveMethod, $object, FormInterface $form): ?string
```
Comment (garak) [inline Form/Type/VichImageType.php]: can we type-hint? E.g. `int $storageResolveMethod, object $object`)

File: Form/Type/VichImageType.php [hunk 6]
```diff
@@ -42,8 +48,18 @@ public function configureOptions(OptionsResolver $resolver): void
...
+            [
+                static::STORAGE_RESOLVE_URI,
+                static::STORAGE_RESOLVE_PATH_RELATIVE,
+                static::STORAGE_RESOLVE_PATH_ABSOLUTE,
+            ]
+        );
+
         $resolver->setAllowedTypes('image_uri', ['bool', 'string', 'callable']);
 
         $imageUriNormalizer = function (Options $options, $imageUri) {
```
Comment (garak) [inline Form/Type/VichImageType.php]: We could add type hint and return type here, too

File: Form/Type/VichImageType.php (hunk 5 above)
Comment (gnutix) [reply Form/Type/VichImageType.php]: I thought about it, but noticed that in `VichFileType` there was a similar method with no type hints. So I restrained, not knowing if there was a good reason for that. Could you confirm we want these type hints ? And should I add them in the other method too while I'm at it?

Comment (garak) [reply Form/Type/VichImageType.php]: I'd prefer such changes in files that need to be changed for other reasons (like this one). We'll handle other files when needed (or in a PR dedicated to bring consistency)

**Comment by gnutix** [comment]
If someone wants to take over this PR, please do. I've removed LiipImagine from my project in the end and don't have more time to dedicate to this.

**Comment by garak** [comment]
Closing in favor of #1045

**Pull request closed without merging**